File: Pro/Assets/Script/Client/Lib/Tool/TransTool.cs
using System;
using System.Text;
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using Object = UnityEngine.Object;

namespace Hello.Game
{
    public static class TransTool
    {
        public static Transform Find(string name, HideFlags flags = HideFlags.None)
        {
            if (string.IsNullOrEmpty(name)) return null;
            GameObject go = GameObject.Find(name);
            if (go == null) go = new GameObject(name);
            go.hideFlags = flags;
            return go.transform;
        }

        public static Transform CreateRoot<T>(HideFlags flags = HideFlags.NotEditable) where T : class
        {
            return CreateRoot(typeof(T).Name, flags);
        }

        public static Transform CreateRoot(Type type,HideFlags flags = HideFlags.NotEditable)
        {
            if (type == null) return null;
            return CreateRoot(type.Name, flags);
        }

        public static Transform CreateRoot(string rootName,HideFlags flags = HideFlags.NotEditable)
        {
            if (string.IsNullOrEmpty(rootName)) return null;
            string name = string.Format("Root<{0}>", rootName);
            Transform root = Find(name, flags);
            Object.DontDestroyOnLoad(root.gameObject);
            return root;
        }

        public static GameObject Find(Transform root,string path,string tip = "")
        {
            if (root == null)
            {
                iTrace.Error("Hello", string.Format("{0} 根节点为空", tip)); return null;
            }
            if (string.IsNullOrEmpty(path))
            {
                iTrace.Error("Hello", string.Format("{0} 路径为空", tip)); return null;
            }
            Transform child = root.Find(path);
            if (child == null)
            {
                iTrace.Error("Hello", string.Format("{0} 根节点:{1},未发现路径为:{2}的子物体", tip, root.name, path)); return null;
            }
            return child.gameObject;
        }

        public static GameObject Find(GameObject root,string path,string tip = "")
        {
            return Find(root.transform, path, tip);
        }

        public static void AddChild(Transform parent,Transform child)
        {
            if (child == null) return;
            if (parent == null) return;
            child.parent = parent;
            child.localScale = Vector3.one;
            child.localPosition = Vector3.zero;
            child.localEulerAngles = Vector3.zero;
        }

        public static void ClearChild(Transform parent,string childName)
        {
            Transform child = parent.Find(childName);
            if (child != null) iTool.Destroy(child.gameObject);
        }

        public static void ClearChildren(Transform parent)
        {
            if (parent == null) return;
            while(parent.childCount != 0)
            {
                Transform tran = parent.GetChild(0);
                Object.DestroyImmediate(tran.gameObject);
            }
        }

        public static void SetChildActive(Transform parent,string path,bool active)
        {
            GameObject child = Find(parent, path);
            if (child != null) child.SetActive(active);
        }

        public static void SetChildrenActive(Transform parent,bool active)
        {
            if (parent == null) return;
            int length = parent.childCount;
            for (int i = 0; i < length; i++)
            {
                Transform child = parent.GetChild(i);
                child.gameObject.SetActive(active);
            }
        }

        public static void RenameChildren(Transform parent,string name)
        {
            if (parent == null) return;
            int length = parent.childCount;
            for (int i = 0; i < length; i++)
            {
                Transform child = parent.GetChild(i);
                child.name = name;
            }
        }

        private static List<string> names = new List<string>();

        public static string GetPath(Transform tran,int beg = 0)
        {
            if (tran == null) return null;
            Transform parent = tran.parent;
            if (parent == null) return tran.name;

            StringBuilder sb = null;
#if UNITY_EDITOR
            bool playing = Application.isPlaying;
            if (!playing)
            {
                sb = new StringBuilder();
            }
            else
#endif
            {
                sb = ObjPool.Instance.Get<StringBuilder>();
            }
            names.Add(tran.name);
            while(parent != null)
            {
                names.Add(parent.name);
                parent = parent.parent;
            }
            if (beg < 0) beg = 0;
            int length = names.Count;
            int start = length - 1 - beg;
            for (int i = start; i > -1 ; --i)
            {
                string name = names[i];
                sb.Append(name);
                if (i > 0) sb.Append("/");
            }
            string path = sb.ToString();
#if UNITY_EDITOR
            if (playing)
#endif
            {
                sb.Remove(0, sb.Length);
                ObjPool.Instance.Add(sb);
            }
            names.Clear();
            return path;
        }

        public static bool IsNull(GameObject go)
        {
            if (go == null) return true;
            if (go.name == "null") return true;
            if (go.Equals(null)) return true;
            return false;
        }

        public static bool IsNull(Transform trans)
        {
            if (trans == null) return true;
            if (trans.name == "null") return true;
            if (trans.Equals(null)) return true;
            return false;
        }

    }
}

File: LuaCode/Tool/TransTool.lua
--[[
 	authors 	:Loong
 	date    	:2017-09-08 15:24:03
 	descrition 	:变换组件工具
--]]

local Gbj = UnityEngine.GameObject

TransTool = {}

local My = TransTool
My.sb = StrBuffer:New{}

--return(GameObject)
--root(Transform):根结点
--path(string):路径
--tip(string):提示
function My.FindChild(root, path, tip)
  local c = My.Find(root, path, tip)
  if c == nil then return nil end
  return c.gameObject
end

--添加子变换
--p(Transform)
--c(Transform)
function My.AddChild(p, c)
  if p == nil then return end
  if c == nil then return end
  c:SetParent(p)
  c.localScale = Vector3.one
  c.localPosition = Vector3.zero
end

--获取指定名称的子变换
--如果不存在则创建
--return(transform)
function My.GetChild(p, cName)
  if p == nil then return end
  local c = p:Find(cName)
  if c then return c end
  local go = Gbj.New()
  go.name = cName
  c = go.transform
  My.AddChild(p, c)
  return c
end

--前置检查
function My.FindCheck(root, path, tip)
  if root == nil then
    iTrace.Error("Loong", tip .. "根结点为空,path:" .. path)
    return false
  end
  if #path < 1 then
    iTrace.Error("Loong", tip .. "路径为空")
    return false;
  end
  do return true end
end

--未发现提示
function My.FindNone(root, path, tip)
  local sb = My.sb
  sb:Apd(tip):Apd(",根节点:"):Apd(root.name);
  sb:Apd(",未发现路径为:"):Apd(path):Apd("的子物体");
  local str = sb:ToStr()
  sb:Dispose()
  iTrace.Error("Loong", str)
end

--return(Transform)
--root(Transform):根结点
--path(string):路径
--tip(string):提示
function My.Find(root, path, tip)
  if not My.FindCheck(root, path, tip) then return end
  local c = root:Find(path)
  if c == nil and tip then
    My.FindNone(root, path, tip)
    return nil
  end
  return c
end

--在根节点下搜索具有指定名称的子节点(包含非直接子节点)
--无特殊需求应该使用Find
--return(Transform)
function My.Search(root, name, tip)
  tip = tip or ""
  if not My.FindCheck(root, name, tip) then return end
  local ty = typeof(UnityEngine.Transform)
  local childs = root:GetComponentsInChildren(ty, true)
  local len = childs.Length - 1
  local child, res = nil, nil
  for i = 0, len do
    child = childs[i]
    if child.name == name then
      res = child
      break
    end
  end
  if res == nil then
    My.FindNone(root, name, tip)
  end
  return res
end

--清理子节点
function My.ClearChildren(p)
  if p == nil then return end
  while p.childCount > 0 do
    local c = p:GetChild(0)
    Gbj.DestroyImmediate(c.gameObject)
  end
end

--设置子游戏对象激活状态
function My.SetChildActive(root, path, active)
  local c = My.Find(root, path)
  active = active or false
  if c then
    c.gameObject:SetActive(active)
  end
end

--重命名子物体
--p(transform):父变换
--newName(string):新名称
--filder(string):子物体有此名称,则跳过
--cat(boolean):子物体激活状态
function My.RenameChildren(p, newName, filter, cat)
  if p == nil then return end
  newName = newName or "none"
  if cat == nil then cat = false end
  local len = p.childCount - 1
  for i = 0, len do
    local c = p:GetChild(i)
    local name = c.name
    if filter ~= name then
      c.name = newName
      c.gameObject:SetActive(cat)
    end
  end
end

--设置所有子物体激活状态
function My.SetChildrenActive(p, at)
  if p == nil then return end
  if at == nil then at = false end
  local len = p.childCount - 1
  for i = 0, len do
    local c = p:GetChild(i)
    c.gameObject:SetActive(at)
  end
end
